File: 0965-univalued-binary-tree/0965-univalued-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isUnivalTree(self, root: Optional[TreeNode]) -> bool:
        self.val = root.val
        self.f = 1

        def dfs(root):
            if not root: return
            dfs(root.left)
            dfs(root.right)
            if root.val != self.val: self.f = 0
        
        dfs(root)
        return self.f == 1

File: 0965-univalued-binary-tree/0965-univalued-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isUnivalTree(TreeNode* root) {
        return check(root, root->val);
    }

    bool check(TreeNode* root, int val) {
        if (root == nullptr) {
            return true;
        }

        return (root->val == val) && check(root->left, val) && check(root->right, val);
    }
};